### PR TITLE
Add OAuth redirect URL to sign-up form.

### PR DESCRIPTION
**Pull request by dvogel**
This adds the redirect URL field to the apply form, which is saved by the changes in this PR: https://github.com/department-of-veterans-affairs/devops/pull/4107

### CODE REVIEW

File: src/containers/ApplyForm.tsx [hunk 1]
```diff
@@ -134,6 +130,18 @@ class ApplyForm extends React.Component<IApplyProps> {
                   <label htmlFor="facilities">VA Facilities API</label>
                 </div>
 
+                <h4> OAuth APIs:</h4>
```
Comment (kalilsn) [inline src/containers/ApplyForm.tsx]:
```suggestion
                <h4>OAuth APIs:</h4>
```

File: src/actions/index.ts [hunk 2]
```diff
@@ -181,11 +187,21 @@ export const submitFormError : ActionCreator<ISubmitFormError> = (status : strin
...
   }
+}
+
+export const validateEmail = (newValue: IErrorableInput) => {
+  validateByPattern(newValue, EMAIL_REGEX, "Must be a valid email address.");
+  return newValue;
+}
+
+export const validateOAuthRedirectURI = (newValue: IErrorableInput) => {
+  const partialUrlPattern = /^http[s]?:[/][/][^/:?]+(:[0-9]+)?([/][^?#]*)?$/;
```
Comment (edpaget) [inline src/actions/index.ts]: I think this is allowing URLs with fragments (like `?test=foo`), which okra will reject. 

```suggestion
try {
  const uri = new URL(newValue.value);
  if (uri.search !== "") {
    newValue.validation = "Cannot include a query fragment";
  }
} catch (error) {
  // This happens when the URL is malformed, for instance does not include a protocol
  newValue.validation = "Must be an http or https URL.";
}
```

Comment (dvogel) [reply src/actions/index.ts]: The regex rejects query strings. It does accept fragments. I'm not tied to that regex but the URL class doesn't provide direct access to the fragment.

Comment (edpaget) [reply src/actions/index.ts]: Hmm I tried it on the linked preview paged and seemed to be able to submit with a query string.

File: src/containers/ApplyForm.tsx [hunk 3]
```diff
@@ -182,6 +192,22 @@ class ApplyForm extends React.Component<IApplyProps> {
     );
   }
 
+  private renderOAuthFields() {
+    if (this.props.inputs.apis.health || this.props.inputs.apis.verification) {
+      const oAuthRedirectURI = this.props.inputs.oAuthRedirectURI;
+      return (
+          <ErrorableTextInput
+            errorMessage={null}
```
Comment (edpaget) [inline src/containers/ApplyForm.tsx]: This should be `this.props.updateOAuthRedirectURI.validation`

File: src/containers/ApplyForm.tsx [hunk 4]
```diff
@@ -195,11 +221,29 @@ class ApplyForm extends React.Component<IApplyProps> {
...
+    return incompleteFields.length === 0;
+  }
+
   private readyToSubmit() {
-    const { inputs: { apis, email, firstName, lastName, organization, termsOfService } } = this.props;
-    return !!email.value && !!firstName.value && !!lastName.value && !!organization.value &&
-           ( apis.verification || apis.health || apis.benefits || apis.facilities ) &&
-           termsOfService;
+    const { inputs: { oAuthRedirectURI, termsOfService } } = this.props;
+    const redirectURIComplete = (!this.anyOAuthApisSelected() || !!oAuthRedirectURI.value);
```
Comment (edpaget) [inline src/containers/ApplyForm.tsx]: Should also check: `!oAuthRedirectURI.validation`

File: src/actions/index.ts (hunk 2 above)
Comment (edpaget) [reply src/actions/index.ts]: Ah got it the issue was elsewhere.

File: src/containers/ApplyForm.tsx (hunk 4 above)
Comment (dvogel) [reply src/containers/ApplyForm.tsx]: Good catch! I put too much trust in my validation test and didn't verify that the submission check was accurate.

Comment (dvogel) [reply src/containers/ApplyForm.tsx]: Fixed.

File: src/containers/ApplyForm.tsx (hunk 3 above)
Comment (dvogel) [reply src/containers/ApplyForm.tsx]: Fixed.

**Pull request merged**